### PR TITLE
[XLA:GPU] Merging amdgpu_executable and nvptx_executable into gpu_executable

### PR DESCRIPTION
**Pull request by jerryyin**
Removed both amdgpu_executable.* and nvptx_executable.*.

Added `absl::variant<std::pair<int, int>, int> compute_capability_` member to cache a common member from either `amd isa version` or `nvdia major minor version number`.

------
Test performed:
//tensorflow/compiler/xla/tests:unary_op_test_gpu compile and pass

### CODE REVIEW

File: tensorflow/compiler/xla/service/gpu/amdgpu_compiler.cc [hunk 1]
```diff
@@ -16,6 +16,7 @@ limitations under the License.
 #include "tensorflow/compiler/xla/service/gpu/amdgpu_compiler.h"
 
 #include <stdlib.h>
+
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/amdgpu_compiler.cc]: remove whiteline

File: tensorflow/compiler/xla/service/gpu/gpu_executable.cc [hunk 2]
```diff
@@ -51,14 +51,17 @@ using tensorflow::tracing::ScopedAnnotation;
 // since we can use timers around thunks.
 GpuExecutable::GpuExecutable(
     const string& text, const std::vector<uint8>& binary,
+    absl::variant<std::pair<int, int>, int> compute_capability,
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/gpu_executable.cc]: I think we may want something like `using GpuVersion = absl::variant<std::pair<int, int>, int>` to make the type more descriptive.

File: tensorflow/compiler/xla/service/gpu/gpu_executable.cc [hunk 3]
```diff
@@ -72,6 +75,42 @@ GpuExecutable::~GpuExecutable() {
                                                assignment_);
 }
 
+Status GpuExecutable::CheckCompatibilityWithServiceExecutableRunOptions(
+    const ServiceExecutableRunOptions* run_options) {
+  se::Stream* main_stream = run_options->stream();
+
+  string vendor = main_stream->parent()->GetDeviceDescription().device_vendor();
+
+  if (vendor == "Advanced Micro Devices, Inc") {
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/gpu_executable.cc]: would it be a good idea to extract these strings from `cuda_gpu_executor.cc` and `rocm_gpu_executor.cc` into a header file?

Comment (jerryyin) [reply tensorflow/compiler/xla/service/gpu/gpu_executable.cc]: Agreed. This thought passed my mind as well. The file being touched then will be on `stream_executor` side. I can file another PR just for the naming concerns and take the opportunity to see if this raw string is used in other places as well.

File: tensorflow/compiler/xla/service/gpu/gpu_executable.cc (hunk 2 above)
Comment (jerryyin) [reply tensorflow/compiler/xla/service/gpu/gpu_executable.cc]: I was trying to follow the [naming from upstream ](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/gpu/gpu_executable.h#L146). I'm good with either of those names.

File: tensorflow/compiler/xla/service/gpu/amdgpu_compiler.cc (hunk 1 above)
Comment (jerryyin) [reply tensorflow/compiler/xla/service/gpu/amdgpu_compiler.cc]: Sounds good

File: tensorflow/compiler/xla/service/gpu/gpu_executable.cc (hunk 3 above)
Comment (whchung) [reply tensorflow/compiler/xla/service/gpu/gpu_executable.cc]: actually use a switch case on `main_stream->parent()->platform_kind()` is even better.

File: tensorflow/compiler/xla/service/gpu/gpu_executable.cc (hunk 2 above)
Comment (whchung) [reply tensorflow/compiler/xla/service/gpu/gpu_executable.cc]: you can see something like:

```
  using BufferAllocToDeviceMemoryMap =
      absl::flat_hash_map<BufferAllocation::Index, se::DeviceMemoryBase>;
```

for types with somewhat complicated semantic, and I believe this particular type suffice to be aliased.

File: tensorflow/compiler/xla/service/gpu/gpu_executable.h [hunk 4]
```diff
@@ -133,10 +135,10 @@ class GpuExecutable : public Executable {
 
   const std::vector<uint8> binary_;
 
-  // The compute capability of the GPU. In the runtime, GpuExecutable check with
-  // either amd's isa version, or nvdia's major minor version, depending on the
-  // gpu hardware
-  absl::variant<std::pair<int, int>, int> compute_capability_;
+  // The GPU version. In the runtime, GpuExecutable check with
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/gpu_executable.h]: comment need re-layout

Comment (jerryyin) [reply tensorflow/compiler/xla/service/gpu/gpu_executable.h]: Do you mean I should update the comment? FYI formatting-wise I have run with clang-format before push.

Comment (whchung) [reply tensorflow/compiler/xla/service/gpu/gpu_executable.h]: clang-format doesn't deal with comments if I understand correctly. Wording here also needs to be tuned. Also the paragraph is missing `.` in its end.

Comment (jerryyin) [reply tensorflow/compiler/xla/service/gpu/gpu_executable.h]: Addressed in latest commit. 

BTW I played around with clang-format a little bit turns out that it does adjust comment as well.

**Pull request merged**